// @mikeyt23/node-cli-utils/dist/esm/DependencyChecker.js
import { spawnSync } from 'node:child_process';
import chalk from 'chalk';
import { platform as rawPlatformString } from 'node:process';
import * as net from 'net';
import { isDockerRunning, isPlatformLinux, isPlatformMac, isPlatformWindows, spawnAsync, whichSync } from './generalUtils.js';
export default class DependencyChecker {
    platformCode;
    constructor() {
        this.platformCode = this.getPlatform();
    }
    hasAllDependencies(dependenciesReport) {
        if (dependenciesReport === null) {
            return false;
        }
        return dependenciesReport.every(entry => entry.value);
    }
    getFormattedReport(report, includePlatform, omitKeys) {
        const platformKey = 'Platform';
        const filteredReport = omitKeys && omitKeys.length > 0 ?
            report.filter(entry => !omitKeys.includes(entry.key)) :
            report;
        const longestKeyLength = Math.max(...filteredReport.map(entry => entry.key.length), platformKey.length);
        let reportString = '\n';
        if (includePlatform) {
            reportString += `${platformKey}${' '.repeat(longestKeyLength - platformKey.length)}: ${this.platformCode}\n`;
        }
        for (const { key, value } of filteredReport) {
            const padding = ' '.repeat(longestKeyLength - key.length);
            reportString += `${key}${padding}: ${value ? chalk.green('true') : chalk.red('false')}\n`;
        }
        return reportString;
    }
    async hasElevatedPermissions() {
        if (this.platformCode === 'win') {
            return await this.winHasElevatedPerms();
        }
        else if (this.platformCode === 'linux') {
            return await this.linuxHasElevatedPerms();
        }
        else if (this.platformCode === 'mac') {
            return await this.linuxHasElevatedPerms();
        }
        return false;
    }
    async winHasElevatedPerms() {
        try {
            await spawnAsync('net', ['session'], { throwOnNonZero: true, stdio: 'ignore' });
            return true;
        }
        catch {
            return false;
        }
    }
    async linuxHasElevatedPerms() {
        if (!process.getuid) {
            throw new Error('Cannot determine if linux user has elevated permissions (process.getuid is undefined)');
        }
        const uid = process.getuid();
        return uid === 0;
    }
    async hasGit() {
        return !!whichSync('git').location;
    }
    async hasDotnetSdkGreaterThanOrEqualTo(minimumMajorVersion) {
        if (!whichSync('dotnet').location) {
            return false;
        }
        const childProc = spawnSync('dotnet', ['--list-sdks'], { encoding: 'utf-8' });
        if (childProc.error) {
            return false;
        }
        const lines = childProc.stdout.split('\n').filter((line) => !!line);
        const lastLine = lines[lines.length - 1];
        let latestMajorVersion;
        try {
            latestMajorVersion = parseInt(lastLine.substring(0, lastLine.indexOf('.')));
        }
        catch {
            throw Error('error parsing results of dotnet --list-sdks');
        }
        return latestMajorVersion >= minimumMajorVersion;
    }
    async hasNodejsGreaterThanOrEqualTo(minimumMajorVersion) {
        if (!whichSync('node').location) {
            return false;
        }
        const childProc = spawnSync('node', ['-v'], { encoding: 'utf-8' });
        if (childProc.error) {
            return false;
        }
        const output = childProc.stdout;
        if (!output || output.length === 0) {
            return false;
        }
        if (!output.startsWith('v')) {
            throw Error('unexpected output for node -v');
        }
        let foundMajorVersion;
        try {
            foundMajorVersion = parseInt(output.substring(1, output.indexOf('.')));
        }
        catch {
            throw Error('error parsing node version');
        }
        return foundMajorVersion >= minimumMajorVersion;
    }
    async hasDocker() {
        return !!whichSync('docker');
    }
    async dockerIsRunning() {
        return await isDockerRunning();
    }
    async hasOpenssl() {
        if (this.platformCode === 'mac') {
            const childProc = spawnSync('brew', ['--prefix', 'openssl'], { encoding: 'utf-8' });
            if (childProc.error) {
                return false;
            }
            const output = childProc.stdout;
            if (!output || output.length === 0) {
                return false;
            }
            return !output.toLowerCase().startsWith('error');
        }
        return !!whichSync('openssl').location;
    }
    async isPortAvailableByEnvKey(envKey) {
        const errorBase = `Cannot lookup port availability for env key ${envKey}`;
        const envVal = process.env[envKey];
        if (!envVal) {
            throw new Error(errorBase + ' - env value not found');
        }
        const port = parseInt(envVal);
        if (isNaN(port)) {
            throw new Error(errorBase + ' - env value could not be parsed into an integer');
        }
        return await this.isPortAvailable(port);
    }
    async isPortAvailable(port) {
        return new Promise((resolve) => {
            const tester = net.connect(port, '127.0.0.1');
            tester.on('connect', () => {
                tester.destroy();
                resolve(false); // port is in use
            });
            tester.on('error', (err) => {
                tester.destroy();
                if (err.code === 'ECONNREFUSED') {
                    resolve(true); // port is available
                }
                else {
                    resolve(false); // some other error occurred, assume port is in use
                }
            });
        });
    }
    getPlatform() {
        if (isPlatformWindows()) {
            return 'win';
        }
        else if (isPlatformMac()) {
            return 'mac';
        }
        else if (isPlatformLinux()) {
            return 'linux';
        }
        else {
            throw Error(`Platform not supported: ${rawPlatformString}. Nodejs process.platform must be win32, darwin or linux.`);
        }
    }
}
