// @mikeyt23/node-cli-utils/dist/cjs/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.yellow = exports.purple = exports.gray = exports.cyan = exports.green = exports.red = exports.color = exports.AnsiColor = void 0;
var AnsiColor;
(function (AnsiColor) {
    AnsiColor["RESET"] = "\u001B[0m";
    AnsiColor["RED"] = "\u001B[31m";
    AnsiColor["GREEN"] = "\u001B[32m";
    AnsiColor["YELLOW"] = "\u001B[33m";
    AnsiColor["CYAN"] = "\u001B[96m";
    AnsiColor["GRAY"] = "\u001B[90m";
    AnsiColor["PURPLE"] = "\u001B[35m";
})(AnsiColor || (exports.AnsiColor = AnsiColor = {}));
const color = (str, colorAnsiCode) => {
    return `${colorAnsiCode}${str}${AnsiColor.RESET}`;
};
exports.color = color;
const red = (str) => (0, exports.color)(str, AnsiColor.RED);
exports.red = red;
const green = (str) => (0, exports.color)(str, AnsiColor.GREEN);
exports.green = green;
const cyan = (str) => (0, exports.color)(str, AnsiColor.CYAN);
exports.cyan = cyan;
const gray = (str) => (0, exports.color)(str, AnsiColor.GRAY);
exports.gray = gray;
const purple = (str) => (0, exports.color)(str, AnsiColor.PURPLE);
exports.purple = purple;
const yellow = (str) => (0, exports.color)(str, AnsiColor.YELLOW);
exports.yellow = yellow;
