// @mikeyt23/node-cli-utils/dist/cjs/DependencyChecker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_child_process_1 = require("node:child_process");
const chalk_1 = __importDefault(require("chalk"));
const node_process_1 = require("node:process");
const net = __importStar(require("net"));
const generalUtils_js_1 = require("./generalUtils.js");
class DependencyChecker {
    constructor() {
        this.platformCode = this.getPlatform();
    }
    hasAllDependencies(dependenciesReport) {
        if (dependenciesReport === null) {
            return false;
        }
        return dependenciesReport.every(entry => entry.value);
    }
    getFormattedReport(report, includePlatform, omitKeys) {
        const platformKey = 'Platform';
        const filteredReport = omitKeys && omitKeys.length > 0 ?
            report.filter(entry => !omitKeys.includes(entry.key)) :
            report;
        const longestKeyLength = Math.max(...filteredReport.map(entry => entry.key.length), platformKey.length);
        let reportString = '\n';
        if (includePlatform) {
            reportString += `${platformKey}${' '.repeat(longestKeyLength - platformKey.length)}: ${this.platformCode}\n`;
        }
        for (const { key, value } of filteredReport) {
            const padding = ' '.repeat(longestKeyLength - key.length);
            reportString += `${key}${padding}: ${value ? chalk_1.default.green('true') : chalk_1.default.red('false')}\n`;
        }
        return reportString;
    }
    async hasElevatedPermissions() {
        if (this.platformCode === 'win') {
            return await this.winHasElevatedPerms();
        }
        else if (this.platformCode === 'linux') {
            return await this.linuxHasElevatedPerms();
        }
        else if (this.platformCode === 'mac') {
            return await this.linuxHasElevatedPerms();
        }
        return false;
    }
    async winHasElevatedPerms() {
        try {
            await (0, generalUtils_js_1.spawnAsync)('net', ['session'], { throwOnNonZero: true, stdio: 'ignore' });
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    async linuxHasElevatedPerms() {
        if (!process.getuid) {
            throw new Error('Cannot determine if linux user has elevated permissions (process.getuid is undefined)');
        }
        const uid = process.getuid();
        return uid === 0;
    }
    async hasGit() {
        return !!(0, generalUtils_js_1.whichSync)('git').location;
    }
    async hasDotnetSdkGreaterThanOrEqualTo(minimumMajorVersion) {
        if (!(0, generalUtils_js_1.whichSync)('dotnet').location) {
            return false;
        }
        const childProc = (0, node_child_process_1.spawnSync)('dotnet', ['--list-sdks'], { encoding: 'utf-8' });
        if (childProc.error) {
            return false;
        }
        const lines = childProc.stdout.split('\n').filter((line) => !!line);
        const lastLine = lines[lines.length - 1];
        let latestMajorVersion;
        try {
            latestMajorVersion = parseInt(lastLine.substring(0, lastLine.indexOf('.')));
        }
        catch (_a) {
            throw Error('error parsing results of dotnet --list-sdks');
        }
        return latestMajorVersion >= minimumMajorVersion;
    }
    async hasNodejsGreaterThanOrEqualTo(minimumMajorVersion) {
        if (!(0, generalUtils_js_1.whichSync)('node').location) {
            return false;
        }
        const childProc = (0, node_child_process_1.spawnSync)('node', ['-v'], { encoding: 'utf-8' });
        if (childProc.error) {
            return false;
        }
        const output = childProc.stdout;
        if (!output || output.length === 0) {
            return false;
        }
        if (!output.startsWith('v')) {
            throw Error('unexpected output for node -v');
        }
        let foundMajorVersion;
        try {
            foundMajorVersion = parseInt(output.substring(1, output.indexOf('.')));
        }
        catch (_a) {
            throw Error('error parsing node version');
        }
        return foundMajorVersion >= minimumMajorVersion;
    }
    async hasDocker() {
        return !!(0, generalUtils_js_1.whichSync)('docker');
    }
    async dockerIsRunning() {
        return await (0, generalUtils_js_1.isDockerRunning)();
    }
    async hasOpenssl() {
        if (this.platformCode === 'mac') {
            const childProc = (0, node_child_process_1.spawnSync)('brew', ['--prefix', 'openssl'], { encoding: 'utf-8' });
            if (childProc.error) {
                return false;
            }
            const output = childProc.stdout;
            if (!output || output.length === 0) {
                return false;
            }
            return !output.toLowerCase().startsWith('error');
        }
        return !!(0, generalUtils_js_1.whichSync)('openssl').location;
    }
    async isPortAvailableByEnvKey(envKey) {
        const errorBase = `Cannot lookup port availability for env key ${envKey}`;
        const envVal = process.env[envKey];
        if (!envVal) {
            throw new Error(errorBase + ' - env value not found');
        }
        const port = parseInt(envVal);
        if (isNaN(port)) {
            throw new Error(errorBase + ' - env value could not be parsed into an integer');
        }
        return await this.isPortAvailable(port);
    }
    async isPortAvailable(port) {
        return new Promise((resolve) => {
            const tester = net.connect(port, '127.0.0.1');
            tester.on('connect', () => {
                tester.destroy();
                resolve(false); // port is in use
            });
            tester.on('error', (err) => {
                tester.destroy();
                if (err.code === 'ECONNREFUSED') {
                    resolve(true); // port is available
                }
                else {
                    resolve(false); // some other error occurred, assume port is in use
                }
            });
        });
    }
    getPlatform() {
        if ((0, generalUtils_js_1.isPlatformWindows)()) {
            return 'win';
        }
        else if ((0, generalUtils_js_1.isPlatformMac)()) {
            return 'mac';
        }
        else if ((0, generalUtils_js_1.isPlatformLinux)()) {
            return 'linux';
        }
        else {
            throw Error(`Platform not supported: ${node_process_1.platform}. Nodejs process.platform must be win32, darwin or linux.`);
        }
    }
}
exports.default = DependencyChecker;
