// @mikeyt23/node-cli-utils/dist/cjs/hostFileUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEolNormalizedWithoutLine = exports.isIpV4Address = exports.changeHostsFile = exports.hostsFileStringHasEntry = exports.hostsFileHasEntry = exports.getHostsFileString = exports.getHostsPath = exports.removeHostsEntry = exports.ensureHostsEntry = void 0;
const promises_1 = __importDefault(require("node:fs/promises"));
const node_os_1 = __importDefault(require("node:os"));
const generalUtils_js_1 = require("./generalUtils.js");
async function ensureHostsEntry(url, ipV4Address = '127.0.0.1') {
    const hostname = (0, generalUtils_js_1.getHostname)(url);
    await changeHostsFile(hostname, ipV4Address, 'add');
}
exports.ensureHostsEntry = ensureHostsEntry;
async function removeHostsEntry(url, ipV4Address = '127.0.0.1') {
    const hostname = (0, generalUtils_js_1.getHostname)(url);
    await changeHostsFile(hostname, ipV4Address, 'remove');
}
exports.removeHostsEntry = removeHostsEntry;
function getHostsPath() {
    return (0, generalUtils_js_1.isPlatformWindows)() ? 'C:/Windows/System32/drivers/etc/hosts' : '/etc/hosts';
}
exports.getHostsPath = getHostsPath;
async function getHostsFileString() {
    return await promises_1.default.readFile(getHostsPath(), { encoding: 'utf-8' });
}
exports.getHostsFileString = getHostsFileString;
async function hostsFileHasEntry(hostname, ipV4Address = '127.0.0.1') {
    const entry = `${ipV4Address} ${hostname}`;
    const hostsFileString = await getHostsFileString();
    return hostsFileStringHasEntry(hostsFileString, entry);
}
exports.hostsFileHasEntry = hostsFileHasEntry;
// Check by normalizing whitespace (collapse consecutive spaces to single spaces) in entry and on each line checked
function hostsFileStringHasEntry(hostsFileString, entry) {
    const normalizedEntry = entry.replace(/\s+/g, ' ');
    const hostsLines = (0, generalUtils_js_1.stringToNonEmptyLines)(hostsFileString).map(l => l.replace(/\s+/g, ' ')).filter(l => !l.startsWith('#'));
    const hasLine = hostsLines.includes(normalizedEntry);
    return hasLine;
}
exports.hostsFileStringHasEntry = hostsFileStringHasEntry;
async function changeHostsFile(hostname, ipV4Address, operation) {
    (0, generalUtils_js_1.requireString)('hostname', hostname);
    if ((0, generalUtils_js_1.hasWhitespace)(hostname)) {
        throw new Error(`Invalid hostname: ${hostname}`);
    }
    if (!isIpV4Address(ipV4Address)) {
        throw new Error(`Invalid ipV4Address: ${ipV4Address}`);
    }
    const isAddition = operation === 'add';
    const isRemoval = operation === 'remove';
    const hostsPath = getHostsPath();
    const entry = `${ipV4Address} ${hostname}`;
    (0, generalUtils_js_1.log)(`checking hosts file: ${hostsPath} for entry ${entry}`);
    const hostsFileString = await getHostsFileString();
    const hasLine = hostsFileStringHasEntry(hostsFileString, entry);
    if (isAddition && hasLine) {
        (0, generalUtils_js_1.log)(`there is an existing entry in the hosts file (${entry}), skipping`);
        return;
    }
    if (isRemoval && !hasLine) {
        (0, generalUtils_js_1.log)(`there is no hosts entry to remove (${entry}), skipping`);
    }
    if (isAddition && !hasLine) {
        (0, generalUtils_js_1.log)('existing entry not found - appending entry to the hosts file');
        await promises_1.default.appendFile(hostsPath, `\n${entry}`);
    }
    if (isRemoval && hasLine) {
        (0, generalUtils_js_1.log)(`existing entry found - removing entry`);
        const hostsWithoutEntry = getEolNormalizedWithoutLine(hostsFileString, entry);
        await promises_1.default.writeFile(hostsPath, hostsWithoutEntry);
    }
}
exports.changeHostsFile = changeHostsFile;
function isIpV4Address(address) {
    const ipV4AddressRegex = /^(\d{1,3}\.){3}\d{1,3}$/;
    if (!ipV4AddressRegex.test(address)) {
        return false;
    }
    const octets = address.split('.').map(Number);
    return octets.every(octet => octet >= 0 && octet <= 255);
}
exports.isIpV4Address = isIpV4Address;
/**
 * The `initialString` will have line endings normalized to use os.EOL and lines with `omitLine` will be removed.
 * Comparisons for which lines should be removed are normalizing whitespace (multiple spaces collapsed into single
 * spaces for the comparison). This is useful to remove instances of a hosts entry, for example.
 * @param initialString The string to normalize remove instances of omitLine from
 * @param omitLine All instance of this string will be omitted from the result
 * @returns A string that has instances of the omitLine removed and all line endings changed to match the os.EOL
 */
function getEolNormalizedWithoutLine(initialString, omitLine) {
    const normalizedOmitLine = omitLine.replace(/\s+/g, ' ');
    const lines = initialString.split('\n')
        .map(l => l.replace(/\r/g, ''))
        .filter(l => l.replace(/\s+/g, ' ') !== normalizedOmitLine);
    return lines.join(node_os_1.default.EOL);
}
exports.getEolNormalizedWithoutLine = getEolNormalizedWithoutLine;
