// @mikeyt23/node-cli-utils/dist/cjs/NodeCliUtilsConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.NodeCliUtilsConfig = void 0;
/**
 * Config to control a few misc settings in the node-cli-utils package. This module exports a singleton instance.
 */
class NodeCliUtilsConfig {
    constructor() {
        this._traceEnabled = false;
        this._orphanProtectionPollingIntervalMillis = 15000;
        this._orphanProtectionLoggingEnabled = false;
        this._orphanProtectionLoggingPath = './orphanProtection.log';
        this._useWslPrefixForDockerCommands = false;
    }
    get traceEnabled() {
        return this._traceEnabled;
    }
    set traceEnabled(value) {
        this._traceEnabled = value;
    }
    get orphanProtectionPollingIntervalMillis() {
        return this._orphanProtectionPollingIntervalMillis;
    }
    set orphanProtectionPollingIntervalMillis(value) {
        this._orphanProtectionPollingIntervalMillis = value;
    }
    get orphanProtectionLoggingEnabled() {
        return this._orphanProtectionLoggingEnabled;
    }
    set orphanProtectionLoggingEnabled(value) {
        this._orphanProtectionLoggingEnabled = value;
    }
    get orphanProtectionLoggingPath() {
        return this._orphanProtectionLoggingPath;
    }
    get useWslPrefixForDockerCommands() {
        return this._useWslPrefixForDockerCommands;
    }
    /**
     * If `true`, docker commands that normally just run `docker` will use `wsl docker` instead. This requires any paths
     * passed to be wsl paths, but these should be handled internally with no additional action required by the caller.
     */
    set useWslPrefixForDockerCommands(value) {
        this._useWslPrefixForDockerCommands = value;
    }
}
exports.NodeCliUtilsConfig = NodeCliUtilsConfig;
/**
 * Singleton instance of {@link NodeCliUtilsConfig}.
 */
exports.config = new NodeCliUtilsConfig();
