// @mikeyt23/node-cli-utils/dist/cjs/TarballUtility.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unpackTarballContents = exports.unpackTarball = exports.createTarball = exports.TarballUtility = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const generalUtils_js_1 = require("./generalUtils.js");
const NodeCliUtilsConfig_js_1 = require("./NodeCliUtilsConfig.js");
/**
 * This utility class exists so we can mock the `which` dependency in unit tests without resorting to libraries that hack the import system.
 */
class TarballUtility {
    constructor(dependencies = {}) {
        var _a, _b, _c, _d;
        /**
         * Creates a gzipped tarball from a directory by spawning a process to run OS-installed `tar` to avoid pulling in npm package dependencies.
         * Note that Windows has tar since Windows 10 1803 (see https://devblogs.microsoft.com/commandline/windows10v1803/.
         *
         * It's possible this isn't 100% reliable due to differences in `tar` versions across platforms. If better normalization
         * is required, consider using the npm package `node-tar` instead.
         * @param directoryToTarball The directory to tarball. The directory name will be used as the root directory in the tarball
         * @param tarballPath The path to the tarball to create - must end with '.tar.gz'
         * @param options See {@link CreateTarballOptions}
         */
        this.createTarball = async (directoryToTarball, tarballPath, options) => {
            (0, generalUtils_js_1.requireValidPath)('directoryToTarball', directoryToTarball);
            (0, generalUtils_js_1.requireString)('tarballPath', tarballPath);
            const defaultOptions = { excludes: [] };
            const mergedOptions = { ...defaultOptions, ...options };
            if (!this.whichSyncFn('tar').location) {
                throw new Error('tar command not found - please install tar on your OS to use this method, or consider using the npm package node-tar instead');
            }
            if (tarballPath.endsWith('.tar.gz') === false) {
                throw new Error(`tarballPath must end with '.tar.gz': ${tarballPath}`);
            }
            const directoryToTarballParentDir = node_path_1.default.dirname(directoryToTarball);
            const directoryToTarballName = node_path_1.default.basename(directoryToTarball);
            const outputDirectory = node_path_1.default.dirname(tarballPath);
            if (!node_fs_1.default.existsSync(outputDirectory)) {
                (0, generalUtils_js_1.trace)(`tarballPath directory does not exist - creating '${outputDirectory}'`);
                await this.mkdirpFn(outputDirectory);
            }
            else if (node_fs_1.default.existsSync(tarballPath)) {
                throw new Error(`tarballPath already exists - delete, move or rename it first: ${tarballPath}`);
            }
            const excludesArgs = mergedOptions.excludes.length > 0 ? mergedOptions.excludes.map(exclude => `--exclude=${exclude}`) : [];
            const verboseFlag = NodeCliUtilsConfig_js_1.config.traceEnabled ? ['-v'] : [];
            const args = [...(verboseFlag), '-czf', tarballPath, '-C', directoryToTarballParentDir, ...excludesArgs, directoryToTarballName];
            const result = await this.spawnAsyncFn('tar', args);
            if (result.code !== 0) {
                throw new Error(`tar command failed with code ${result.code}`);
            }
            (0, generalUtils_js_1.trace)('tarball created: ' + tarballPath);
        };
        /**
         * Unpacks a gzipped tarball by spawning a process to run OS-installed `tar` to avoid pulling in npm package dependencies.
         * This method will throw an error if the unpackDirectory is not empty, unless the `throwOnNonEmptyUnpackDir` option is set to false.
         * @param tarballPath The path to the tarball to unpack
         * @param unpackDirectory The directory to unpack the tarball into
         * @param options The options to use when unpacking the tarball. See {@link TarballUnpackOptions}.
         */
        this.unpackTarball = async (tarballPath, unpackDirectory, options) => {
            (0, generalUtils_js_1.requireValidPath)('tarballPath', tarballPath);
            (0, generalUtils_js_1.requireString)('unpackDirectory', unpackDirectory);
            if (!this.whichSyncFn('tar').location) {
                throw new Error('tar command not found - please install tar on your OS to use this method, or consider using the npm package node-tar instead');
            }
            const defaultOptions = { createDirIfNotExists: false, stripComponents: 0, throwOnNonEmptyUnpackDir: true };
            const mergedOptions = { ...defaultOptions, ...options };
            if (mergedOptions.stripComponents < 0) {
                throw new Error('stripComponents must be greater than or equal to 0 if provided');
            }
            const unpackedDirExists = node_fs_1.default.existsSync(unpackDirectory);
            if (unpackedDirExists && !this.isDirectory(unpackDirectory)) {
                throw new Error(`unpackDirectory exists but is not a directory: ${unpackDirectory}`);
            }
            if (mergedOptions.createDirIfNotExists && !unpackedDirExists) {
                await this.tryCreateDirectory(unpackDirectory);
            }
            if (!mergedOptions.createDirIfNotExists && !unpackedDirExists) {
                throw new Error(`unpackDirectory does not exist: ${unpackDirectory}`);
            }
            if (mergedOptions.throwOnNonEmptyUnpackDir && this.dirIsNotEmpty(unpackDirectory)) {
                throw new Error(`unpackDirectory exists but is not empty: ${unpackDirectory}`);
            }
            const verboseFlag = NodeCliUtilsConfig_js_1.config.traceEnabled ? ['-v'] : [];
            const args = [...(verboseFlag), '-xzf', tarballPath, '-C', unpackDirectory, '--strip-components', mergedOptions.stripComponents.toString()];
            const result = await this.spawnAsyncFn('tar', args);
            if (result.code !== 0) {
                throw new Error(`tar command failed with code ${result.code}`);
            }
            (0, generalUtils_js_1.trace)(`tarball unpacked at ${unpackDirectory}`);
        };
        /**
         * A more opinionated version of {@link unpackTarball} that assumes you want to create the directory and strip the first directory out of the unpacked files.
         * @param tarballPath The path to the tarball to unpack
         * @param unpackDirectory The directory to unpack the tarball into - will be created if it doesn't exist and will throw if it exists but is not empty
         * @param stripComponents The number of leading directories to strip out of the unpacked files, defaults to 1
         */
        this.unpackTarballContents = async (tarballPath, unpackDirectory, stripComponents = 1) => {
            await this.unpackTarball(tarballPath, unpackDirectory, { stripComponents, createDirIfNotExists: true });
        };
        this.isDirectory = (path) => {
            try {
                const stats = this.statSyncFn(path);
                return stats.isDirectory();
            }
            catch (err) {
                (0, generalUtils_js_1.trace)('error checking idDirectory (returning false)', err);
                return false;
            }
        };
        this.dirIsNotEmpty = (dirPath) => {
            try {
                const stats = this.statSyncFn(dirPath);
                return stats.isDirectory() && node_fs_1.default.readdirSync(dirPath).length > 0;
            }
            catch (err) {
                throw new generalUtils_js_1.ExtendedError('Error checking dirIsNotEmpty - see innerError', err);
            }
        };
        this.tryCreateDirectory = async (dirPath) => {
            try {
                await this.mkdirpFn(dirPath);
            }
            catch (err) {
                throw new generalUtils_js_1.ExtendedError('Error creating unpackDirectory - see innerError', err);
            }
        };
        this.whichSyncFn = (_a = dependencies.whichSyncFn) !== null && _a !== void 0 ? _a : generalUtils_js_1.whichSync;
        this.spawnAsyncFn = (_b = dependencies.spawnAsyncFn) !== null && _b !== void 0 ? _b : generalUtils_js_1.spawnAsync;
        this.statSyncFn = (_c = dependencies.statSyncFn) !== null && _c !== void 0 ? _c : node_fs_1.default.statSync;
        this.mkdirpFn = (_d = dependencies.mkdirpFn) !== null && _d !== void 0 ? _d : generalUtils_js_1.mkdirp;
    }
}
exports.TarballUtility = TarballUtility;
const defaultUtil = new TarballUtility();
exports.createTarball = defaultUtil.createTarball;
exports.unpackTarball = defaultUtil.unpackTarball;
exports.unpackTarballContents = defaultUtil.unpackTarballContents;
