// @mikeyt23/node-cli-utils/dist/cjs/testUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertErrorMessageEquals = exports.assertErrorMessageIncludes = exports.assertErrorMessageStartsWith = exports.fileExistsAndIsNonZero = exports.ensureEmptyTempDir = exports.only = exports.fixturesDir = exports.tempDir = void 0;
const node_assert_1 = __importDefault(require("node:assert"));
const node_fs_1 = __importDefault(require("node:fs"));
const promises_1 = __importDefault(require("node:fs/promises"));
exports.tempDir = './test/temp';
exports.fixturesDir = './test/fixtures';
exports.only = { only: true }; // Convenience object to make it easier to mark tests as "only"
const ensureEmptyTempDir = async (dir) => {
    if (!dir) {
        throw new Error('dir is required');
    }
    if (!dir.startsWith('test/temp') && !dir.startsWith('test\\temp') && !dir.startsWith('./test/temp') && !dir.startsWith('.\\test\\temp')) {
        throw new Error(`dir must start with 'test/temp': ${dir}`);
    }
    if (node_fs_1.default.existsSync(dir)) {
        await promises_1.default.rm(dir, { recursive: true, force: true });
    }
    await promises_1.default.mkdir(dir, { recursive: true });
};
exports.ensureEmptyTempDir = ensureEmptyTempDir;
function fileExistsAndIsNonZero(filePath) {
    try {
        const stats = node_fs_1.default.statSync(filePath);
        return stats.isFile() && stats.size > 0;
    }
    catch (err) {
        const e = err;
        if (e.code === 'ENOENT') {
            return false; // File doesn't exist
        }
        throw err; // Rethrow other errors
    }
}
exports.fileExistsAndIsNonZero = fileExistsAndIsNonZero;
function assertErrorMessageStartsWith(err, expectedStartsWith) {
    (0, node_assert_1.default)(err instanceof Error);
    node_assert_1.default.strictEqual(err.message.startsWith(expectedStartsWith), true, `Error message did not start with expected value. Actual: "${err.message}".`);
    return true;
}
exports.assertErrorMessageStartsWith = assertErrorMessageStartsWith;
function assertErrorMessageIncludes(err, expectedIncludes) {
    (0, node_assert_1.default)(err instanceof Error);
    node_assert_1.default.strictEqual(err.message.includes(expectedIncludes), true, `Error message did not include the expected value. Actual: "${err.message}".`);
    return true;
}
exports.assertErrorMessageIncludes = assertErrorMessageIncludes;
function assertErrorMessageEquals(err, expected) {
    (0, node_assert_1.default)(err instanceof Error);
    node_assert_1.default.strictEqual(err.message, expected, `Error message did not equal the expected value. Actual: "${err.message}".`);
    return true;
}
exports.assertErrorMessageEquals = assertErrorMessageEquals;
